### DISCUSSION TITLE
How get position of element with specific class

### DISCUSSION THREAD

**Discussion by abombelli** [dompdf/dompdf#3378] [Q&A] [answered]
Is there a way to get the x/y position of an element with a specific class in the rendered PDF?
We would like to draw some lines to connect content placed in table cells. As they grow according to the content in it, we need a way to get the position of the DIV's in this cells.
Any suggested approach?

**Comment by bsweeney** [comment]
You can get the border box of a frame after rendering through the `end_frame` callback. From there you can get the coordinates you need to draw a line. Here's a simple example:

```
$elements = [];
$dompdf = new Dompdf();
$dompdf->loadHtml('
    <div id="box1" style="position: absolute; inset: 0 auto auto 0; height: 100px; width: 100px; background-color: #00ff0033"></div>
    <div id="box2" style="position: absolute; inset: 300 0 auto auto; height: 100px; width: 100px; background-color: #00ff0033"></div>
');
$dompdf->setCallbacks([
    [
        'event' => 'end_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$elements) {
            $node = $frame->get_node();
            if ($node->hasAttributes() && ($id = $node->attributes->getNamedItem("id")) !== null && \in_array($id->nodeValue, ["box1", "box2"], true)) {
                $elements[$id->nodeValue] = $frame->get_border_box();
            }
        }
    ]
]);
$dompdf->render();

$canvas = $dompdf->getCanvas();
$color = \Dompdf\Css\Color::parse("#009900");
$x1 = $elements["box1"]["x"] + $elements["box1"]["w"];
$y1 = $elements["box1"]["y"] + ($elements["box1"]["h"] / 2);
$x2 = $elements["box2"]["x"];
$y2 = $elements["box2"]["y"] + ($elements["box2"]["h"] / 2);
$canvas->line($x1, $y1, $x2, $y2, $color, 3.0);
```

I made some assumptions for this simple example: single page; known box arrangement. The border box should give you enough information to determine line coordinates if the boxes are positioned dynamically during rendering, even if it's across pages.

Things get a little more complicated if you need to render the line _behind_ other page contents. Because PDFs layer elements in order of rendering, the line will appear on top of the content of the page using this method. I think to be able to render the line in this way but under other elements an enhancement would have to be built to allow specifying the rendering order of the stack within Cpdf.

**Reply by abombelli** [reply]
Thank you so much. Works perfect.

As for the render order. If this is possible, to have control over that would be great.
Would it be an approach to store the initial rendered content and place/load it then over the rendered lines?

As for the lines; is there also a function to draw some multipoint shapes?
Or could I pass an SVG element to $canvas->image ?

**Reply by bsweeney** [reply]
You definitely could capture the boxes and then write them out later, so long as everything is on a single page. I rewrote the sample to do it this way.

```
$dompdf = new Dompdf();
$dompdf->loadHtml('
    <div id="box1" style="position: absolute; inset: 0 auto auto 0; height: 100px; width: 100px; background-color: #006600;"></div>
    <div id="box2" style="position: absolute; inset: 300 0 auto auto; height: 100px; width: 100px; background-color: #006600;"></div>
');
$coords = [];
$objectRef;
$dompdf->setCallbacks([
    [
        'event' => 'begin_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$objectRef) {
            $node = $frame->get_node();
            if ($node->hasAttributes() && ($id = $node->attributes->getNamedItem("id")) !== null && $id->nodeValue === "box1") {
                $objectRef = $canvas->open_object();
            }
        }
    ],
    [
        'event' => 'end_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$coords, &$objectRef) {
            $node = $frame->get_node();
            if ($node->hasAttributes() && ($id = $node->attributes->getNamedItem("id")) !== null && \in_array($id->nodeValue, ["box1", "box2"], true)) {
                $coords[$id->nodeValue] = $frame->get_border_box();

                if ($id->nodeValue == "box2") {
                    $canvas->close_object($objectRef);

                    $color = \Dompdf\Css\Color::parse("#009900cc");
                    $x1 = $coords["box1"]["x"] + ($coords["box1"]["w"] / 2);
                    $y1 = $coords["box1"]["y"] + ($coords["box1"]["h"] / 2);
                    $x2 = $coords["box2"]["x"] + ($coords["box2"]["w"] / 2);
                    $y2 = $coords["box2"]["y"] + ($coords["box2"]["h"] / 2);
                    $canvas->line($x1, $y1, $x2, $y2, $color, 3.0, [], "round");

                    $canvas->add_object($objectRef, "add");
                }
            }
        }
    ]
]);
```

FYI the previous method I tried for rendering the line behind the boxes otherwise was to add an empty object before rendering the boxes, close it, continue rendering, then reopen the object later to add the line. I would have expected that to render the line under the boxes, but it did not. Maybe something in my technique was flawed, I'd have to investigate more to know for sure.

To draw a multi-point shape you would have to do a bit of work. Some of the necessary functionality isn't available in the [Canvas](https://github.com/dompdf/dompdf/blob/master/src/Canvas.php), the current implementation of which supports the following: line, arc, circle, rectangle, and polygon. To draw a multi-point shape, you would have to draw each segment manually. You can get access to the full backend functionality by accessing it directly (e.g., with Cpdf `$cpdf = $dompdf->getCanvas()->get_cpdf();`).

Depending on the complexity of the shape you might consider generating an SVG document that uses a path and then inserting that into the document.

**Reply by abombelli** [reply]
Thanks a lot.
$canvas->polygon did it!

As for the render order, this new approach works!
Now I'm trying to make all this dynamic. I have an array with source and target id's of boxes to connect.
When adding new callbacks, only the las line is drawn.

Could you point me to the right direction?
I have an array like `$connectors = [["box1", "box2"], ["box8", "box6"]];` to draw the lines.
My first thought was to loop through the array in begin_frame and end_frame, but then I override `$objectRef` on each itineration.
Or is there a better approach?

**Reply by bsweeney** [reply]
If all the boxes are rendered at the same time the easiest way to handle that would be to add your open/close object calls on a wrapper element. Something like this:

```
<div id="box_container">
   <div id="box1"></div>
</div>
```

You could also try to do something more complex like use a separate object for each grouping (`$connectors = ["elements" => ["box1" => null, "box2" => null], $objRef => null]`), but the logic gets a lot more complex.

**Reply by abombelli** [reply]
I tried to recreate your suggestion, but I'm not so sure on how to incorporate it into the `$objectRef;` example.
Do I need to check in begin_frame and end_frame for box_container?

**Reply by bsweeney** [reply, marked as answer]
There are a lot of ways you could try to achieve the desired result. Here's a somewhat complex but generalized example.

```
$dompdf = new Dompdf();
$dompdf->loadHtml('
    <div id="box-container">
        <div id="box1" style="position: absolute; inset: 0 auto auto 0; height: 100px; width: 100px; background-color: #006600;"></div>
        <div id="box2" style="position: absolute; inset: 300 0 auto auto; height: 100px; width: 100px; background-color: #006600;"></div>
        <div id="box3" style="position: absolute; inset: 500 auto auto 100; height: 100px; width: 100px; background-color: #000066;"></div>
        <div id="box4" style="position: absolute; inset: 100 75 auto auto; height: 100px; width: 100px; background-color: #000066;"></div>
    </div>
');
$connectors = [
    "elements" => [],
    "connections" => [
        ["box1", "box2"],
        ["box4", "box3"]
    ]
];
$objectRef = null;
$dompdf->setCallbacks([
    [
        'event' => 'begin_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$objectRef) {
            $node = $frame->get_node();
            if ($node->hasAttributes() && ($id = $node->attributes->getNamedItem("id")) !== null && $id->nodeValue === "box-container") {
                $objectRef = $canvas->open_object();
            }
        }
    ],
    [
        'event' => 'end_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$connectors) {
            $node = $frame->get_node();

            $parent = $frame->get_parent();
            $parent_node = $parent->get_node();
            $parent_id = ($node->hasAttributes() && $parent_node->attributes->getNamedItem("id") !== null) ? $parent_node->attributes->getNamedItem("id") : null;
            if (empty($parent_id) || $parent_id->nodeValue !== "box-container") {
                return;
            }

            $id = ($node->hasAttributes() && $node->attributes->getNamedItem("id") !== null) ? $node->attributes->getNamedItem("id") : null;
            if (empty($id)) {
                return;
            }
            
            $connectors["elements"][$id->nodeValue] = [
                "bbox" => $frame->get_border_box(),
                "color" => $frame->get_style()->background_color
            ];
        }
    ],
    [
        'event' => 'end_frame',
        'f' => function ($frame, $canvas, $fontMetrics) use (&$connectors, &$objectRef) {
            $node = $frame->get_node();
            $id = ($node->hasAttributes() && $node->attributes->getNamedItem("id") !== null) ? $node->attributes->getNamedItem("id") : null;
            if (empty($id) || $id->nodeValue !== "box-container") {
                return;
            }

            $canvas->close_object($objectRef);

            $elements = $connectors["elements"];
            foreach ($connectors["connections"] as $boxes) {
                $color = $elements[$boxes[0]]["color"];
                $color["alpha"] = .6;
                $x1 = $elements[$boxes[0]]["bbox"]["x"] + ($elements[$boxes[0]]["bbox"]["w"] / 2);
                $y1 = $elements[$boxes[0]]["bbox"]["y"] + ($elements[$boxes[0]]["bbox"]["h"] / 2);
                $x2 = $elements[$boxes[1]]["bbox"]["x"] + ($elements[$boxes[1]]["bbox"]["w"] / 2);
                $y2 = $elements[$boxes[1]]["bbox"]["y"] + ($elements[$boxes[1]]["bbox"]["h"] / 2);
                $canvas->line($x1, $y1, $x2, $y2, $color, 3.0, [], "round");
            }

            $canvas->add_object($objectRef, "add");
        }
    ]
]);
$dompdf->render();
$dompdf->stream();
```

**Reply by abombelli** [reply]
Sorry about the late feedback.
Thanks a lot for the input!
